src/control/lib/hardware/pciutils/bindings.go

```diff
@@ -0,0 +1,120 @@
+//
+// (C) Copyright 2024 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package pciutils
+
+import (
+	"os"
+	"unsafe"
+
+	"github.com/daos-stack/daos/src/control/lib/hardware"
+	"github.com/pkg/errors"
+)
+
+/*
+#cgo LDFLAGS: -lpci
+
+#include <stdlib.h>
+#include <pci/pci.h>
+*/
+import "C"
+
+type api struct{}
+
+func (api *api) Cleanup() {}
+
+func speedToFloat(speed uint16) float32 {
+	var mant float32
+
+	switch speed {
+	case 1:
+		mant = 2.5
+	case 2:
+		mant = 5
+	case 3:
+		mant = 8
+	case 4:
+		mant = 16
+	case 5:
+		mant = 32
+	case 6:
+		mant = 64
+	default:
+		return 0
+	}
+
+	return mant * 1e+9
+}
+
+func (api *api) PCIeCapsFromConfig(cfgBytes []byte, dev *hardware.PCIDevice) error {
+	if len(cfgBytes) == 0 {
+		return errors.New("empty config")
+	}
+	if dev == nil {
+		return errors.New("nil dev")
+	}
+
+	access := C.pci_alloc()
+	if access == nil {
+		return errors.New("pci_alloc() failed")
+	}
+	access.method = C.PCI_ACCESS_DUMP
+
+	tmpFile, err := os.CreateTemp("", "pciutils")
+	if err != nil {
+		return err
+	}
+	fName := tmpFile.Name()
+	defer os.Remove(fName)
+
+	cfgBytes = append([]byte("01:00.0 device #1\n"), cfgBytes...)
```

Comment:
This is a little confusing to me. Why do we need to tack this on? Are this address and device number always right?

Reply:
This preamble is expected by the library when reading config from file (lspci "-F" option). The actual address is irrelevant because we are just extracting from the config content, we don't read the address. Without an address and device values in the preamble, the library will refuse to parse the file content.

Reply:
I've decided that the preamble should be applied to the config content before calling into the library. The library itself shouldn't be concerned with altering the config bytes input. The preamble is now applied in the server code before calling into the library. I've added some more comments to improve clarity.